pkg/services/ngalert/state/manager.go

```diff
@@ -113,69 +113,97 @@ func (st *Manager) Warm(ctx context.Context, rulesReader RuleReader) {
 
 	statesCount := 0
 	states := make(map[int64]map[string]*ruleStates, len(orgIds))
-	for _, orgId := range orgIds {
-		// Get Rules
-		ruleCmd := ngModels.ListAlertRulesQuery{
-			OrgID: orgId,
-		}
-		alertRules, err := rulesReader.ListAlertRules(ctx, &ruleCmd)
-		if err != nil {
-			st.log.Error("Unable to fetch previous state", "error", err)
-		}
+	for _, orgID := range orgIds {
+		orgStates, count := st.warmOrg(ctx, orgID, rules)
+		statesCount += count
+		states[orgID] = orgStates
+	}
+	st.cache.setAllStates(states)
+	st.log.Info("State cache has been initialized", "states", statesCount, "duration", time.Since(startTime))
+}
 
-		ruleByUID := make(map[string]*ngModels.AlertRule, len(alertRules))
-		for _, rule := range alertRules {
-			ruleByUID[rule.UID] = rule
-		}
+func (st *Manager) warmOrg(ctx context.Context, orgID int64, rules RuleReader) (map[string]*ruleStates, int) {
+	// Get Rules
+	ruleCmd := ngModels.ListAlertRulesQuery{
+		OrgID: orgID,
+	}
+	alertRules, err := rules.ListAlertRules(ctx, &ruleCmd)
+	if err != nil {
+		st.log.Error("Unable to fetch previous state", "error", err)
+	}
+
+	ruleByUID := make(map[string]*ngModels.AlertRule, len(alertRules))
+	for _, rule := range alertRules {
+		ruleByUID[rule.UID] = rule
+	}
+
+	states := make(map[string]*ruleStates, len(ruleByUID))
+	count := 0
+
+	// Get Instances
+	cmd := ngModels.ListAlertInstancesQuery{
+		RuleOrgID: orgID,
+	}
+	alertInstances, err := st.instanceStore.ListAlertInstances(ctx, &cmd)
+	if err != nil {
+		st.log.Error("Unable to fetch previous state", "error", err)
+	}
 
-		orgStates := make(map[string]*ruleStates, len(ruleByUID))
-		states[orgId] = orgStates
+	orphanedInstances := make([]ngModels.AlertInstanceKey, 0)
```

Comment:
Would it not be easier (and perhaps faster too) to do something like this in SQL:

```
DELETE FROM alert_instance WHERE rule_uid NOT IN (SELECT uid FROM alert_rule)
```

and then load the remaining instances?

Reply:
I moved to a different implementation that's closer to this, IMO we shouldn't do this as part of warm() but in a separate phase that can be invoked on its own.

I couldn't do this exactly though, since rule UIDs aren't unique, and this needs to work across multiple orgs, so it needs to take the (org, UID) tuple into account rather than just comparing UIDs.

A multi-column `WHERE NOT IN` statement is not part of the SQL standard, it's an extension, unfortunately. It works in Postgres, but is not guaranteed to work in other databases.

I've done the next best thing to this - the new implementation loads only the distinct rule UIDs and orgIDs from the instances, not the whole thing. This query can be answered purely by an index and without a table scan, so it's _much_ faster than the cache warm.